src/irmin/proof.ml

```diff
@@ -139,7 +140,24 @@ struct
       let set = Hashes.create 13 in
       let singleton_inodes = Hashes.create 13 in
       let rev_elts = ref [] in
-      Produce { set; singleton_inodes; rev_elts }
+      let rev_elts_size = ref 0 in
+      Produce { set; singleton_inodes; rev_elts; rev_elts_size }
+
+    let list_insert l idx v =
+      (* [list_insert l 0 v] is [v :: l] *)
+      assert (idx >= 0);
+      let rec aux l i acc =
+        if i = 0 then List.rev_append acc (v :: l)
+        else
+          match l with
+          | [] -> failwith "list_insert: input list too short"
+          | hd :: tl -> aux tl (i - 1) (hd :: acc)
+      in
+      aux l idx []
```

Comment:
I think it's reasonable to put this in `import.ml` as `List.insert_exn`, but no real need to.

Reply:
I'll do that in the forward port